Libraries/LibLine/SuggestionManager.cpp

```diff
@@ -0,0 +1,190 @@
+        m_largest_common_suggestion_prefix_length = common_suggestion_prefix;
+    } else {
+        m_largest_common_suggestion_prefix_length = 0;
+    }
+}
+
+void SuggestionManager::next()
+{
+    if (m_suggestions.size())
+        m_next_suggestion_index = (m_next_suggestion_index + 1) % m_suggestions.size();
+    else
+        m_next_suggestion_index = 0;
+}
+
+void SuggestionManager::previous()
+{
+    if (m_next_suggestion_index == 0)
+        m_next_suggestion_index = m_suggestions.size();
+    m_next_suggestion_index--;
+}
+
+const CompletionSuggestion& SuggestionManager::suggest()
+{
+    m_last_shown_suggestion = m_suggestions[m_next_suggestion_index];
+    m_selected_suggestion_index = m_next_suggestion_index;
+    return m_last_shown_suggestion;
+}
+
+void SuggestionManager::set_current_suggestion_initiation_index(size_t index)
+{
+
+    if (m_last_shown_suggestion_display_length)
+        m_last_shown_suggestion.start_index = index - m_next_suggestion_static_offset - m_last_shown_suggestion_display_length;
+    else
+        m_last_shown_suggestion.start_index = index - m_next_suggestion_static_offset - m_next_suggestion_invariant_offset;
+
+    m_last_shown_suggestion_display_length = m_last_shown_suggestion.text_view.length();
+    m_last_shown_suggestion_was_complete = true;
+}
+
+SuggestionManager::CompletionAttemptResult SuggestionManager::attempt_completion(CompletionMode mode, size_t initiation_start_index)
+{
+    CompletionAttemptResult result { mode };
+
+    if (m_next_suggestion_index < m_suggestions.size()) {
+        auto can_complete = m_next_suggestion_invariant_offset <= m_largest_common_suggestion_prefix_length;
+        if (!m_last_shown_suggestion.text.is_null()) {
+            ssize_t actual_offset;
+            size_t shown_length = m_last_shown_suggestion_display_length;
+            switch (mode) {
+            case CompletePrefix:
+                actual_offset = 0;
+                break;
+            case ShowSuggestions:
+                actual_offset = 0 - m_largest_common_suggestion_prefix_length + m_next_suggestion_invariant_offset;
+                if (can_complete)
+                    shown_length = m_largest_common_suggestion_prefix_length + m_last_shown_suggestion.trivia_view.length();
+                break;
+            default:
+                if (m_last_shown_suggestion_display_length == 0)
+                    actual_offset = 0;
+                else
+                    actual_offset = 0 - m_last_shown_suggestion_display_length + m_next_suggestion_invariant_offset;
+                break;
+            }
+
+            result.offset_region_to_remove = { m_next_suggestion_invariant_offset, shown_length };
+            result.new_cursor_offset = actual_offset;
+        }
+
+        auto& suggestion = suggest();
+        set_current_suggestion_initiation_index(initiation_start_index);
+
+        if (mode == CompletePrefix) {
+            // only auto-complete *if possible*
```

Comment:
https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#comments

Reply:
uhhh...capitalisation?

Reply:
Mind if I do this in a separate commit?
tracking the comments through the history is a bit...annoying.
I could do that if needed though.

Reply:
That's fine! :)

Reply:
Github seems very broken today...

anyway, proper comments go _brrrrrrrrrrrrrrrrrrrrrrrrrrrrrr_